ext/SummationByPartsOperatorsOptimForwardDiffExt.jl

```diff
@@ -0,0 +1,222 @@
+    V = vandermonde_matrix(basis_functions_orthonormalized, nodes)
+    V_x = vandermonde_matrix(basis_functions_orthonormalized_derivatives, nodes)
+    # Here, W satisfies W'*W = I
+    # W = [V; -V_x]
+
+    B = spzeros(eltype(nodes), N, N)
+    B[1, 1] = -1
+    B[N, N] = 1
+
+    R = B * V / 2
+    x_length = x_max - x_min
+    S = zeros(eltype(nodes), N, N)
+    A = zeros(eltype(nodes), N, K)
+    SV = zeros(eltype(nodes), N, K)
+    PV_x = zeros(eltype(nodes), N, K)
+    S_cache = DiffCache(S)
+    A_cache = DiffCache(A)
+    SV_cache = DiffCache(SV)
+    PV_x_cache = DiffCache(PV_x)
+    daij_dsigmak = zeros(eltype(nodes), N, K, L)
+    daij_drhok = zeros(eltype(nodes), N, K, N)
+    p = (V, V_x, R, x_length, S_cache, A_cache, SV_cache, PV_x_cache, daij_dsigmak, daij_drhok)
+
+    x0 = zeros(L + N)
+    fg!(F, G, x) = optimization_function_and_grad!(F, G, x, p)
+    result = optimize(Optim.only_fg!(fg!), x0, LBFGS(), options)
+
+    x = minimizer(result)
+    sigma = x[1:L]
+    rho = x[(L + 1):end]
+    S = create_S(sigma, N)
+    P = create_P(rho, x_length)
+    weights = diag(P)
+    Q = S + B/2
+    D = inv(P) * Q
+    return weights, D
+end
+
+@views function optimization_function_and_grad!(F, G, x, p)
+    V, V_x, R, x_length, S_cache, A_cache, SV_cache, PV_x_cache, daij_dsigmak, daij_drhok = p
+    S = get_tmp(S_cache, x)
+    A = get_tmp(A_cache, x)
+    SV = get_tmp(SV_cache, x)
+    PV_x = get_tmp(PV_x_cache, x)
+    (N, _) = size(R)
+    L = div(N * (N - 1), 2)
+    sigma = x[1:L]
+    rho = x[(L + 1):end]
+    set_S!(S, sigma, N)
+    P = create_P(rho, x_length)
+    mul!(SV, S, V)
+    mul!(PV_x, P, V_x)
+    @. A = SV - PV_x + R
+    if !isnothing(G)
+        fill!(daij_dsigmak, zero(eltype(daij_dsigmak)))
+        for k in axes(daij_dsigmak, 3)
+            for j in axes(daij_dsigmak, 2)
+                for i in axes(daij_dsigmak, 1)
+                    l_tilde = k + i - N * (i - 1) + div(i * (i - 1), 2)
+                    # same as above, but needs more type conversions
+                    # l_tilde = Int(k + i - (i - 1) * (N - i/2))
+                    if i + 1 <= l_tilde <= N
+                        daij_dsigmak[i, j, k] += V[l_tilde, j]
+                    else
+                        C = N^2 - 3*N + 2*i - 2*k + 1/4
+                        if C >= 0
+                            D = sqrt(C)
+                            D_plus_one_half = D + 0.5
+                            D_plus_one_half_trunc = trunc(D_plus_one_half)
+                            if D_plus_one_half == D_plus_one_half_trunc
+                                int_D_plus_one_half = trunc(Int, D_plus_one_half_trunc)
+                                l_hat1 = N + int_D_plus_one_half - 1
+                                l_hat2 = N - int_D_plus_one_half
+                                if 1 <= l_hat1 <= i - 1
+                                    daij_dsigmak[i, j, k] -= V[l_hat1, j]
```

Comment:
This line is reported as not being covered by tests. Any idea why? Is it needed or is this case impossible?

Reply:
Yes, this case is indeed impossible: We have $\hat{l}_1 = N - 1/2 + \sqrt{C} > N - 1 \ge i - 1$. Thanks for the hint! See [fd470ee](https://github.com/ranocha/SummationByPartsOperators.jl/pull/268/commits/fd470ee623f3a9585a5be3c6558fd2e99cf9af98).

Reply:
Thanks